test/unit/utils/test-size-observer.js

```diff
@@ -0,0 +1,179 @@
+      }
+      return observer;
+    });
+  });
+
+  function createCallbackCaller() {
+    const results = [];
+    let deferred = null;
+    const caller = (value) => {
+      console.log('QQQQ: result:', value);
+      results.push(value);
+      if (deferred) {
+        const next = results.shift();
+        deferred.resolve(next);
+        deferred = null;
+      }
+    };
+    caller.next = () => {
+      if (results.length > 0) {
+        const next = results.shift();
+        return Promise.resolve(next);
+      }
+      deferred = new Deferred();
+      return deferred.promise;
+    };
+    caller.isEmpty = () => results.length == 0;
+    return caller;
+  }
+
+  it('should measure content size', async () => {
+    const promise = measureContentSize(element);
+    observer.notify([
+      {target: element, contentRect: {width: 101, height: 102}},
+    ]);
+    const {width, height} = await promise;
+    expect(width).to.equal(101);
+    expect(height).to.equal(102);
+    expect(observer.elements).to.not.include(element);
+  });
+
+  it('should observe changes', async () => {
+    const callbackCaller = createCallbackCaller();
+    observeContentSize(element, callbackCaller);
+
+    // First response.
+    observer.notify([
+      {target: element, contentRect: {width: 101, height: 102}},
+    ]);
+    const size1 = await callbackCaller.next();
+    expect(size1.width).to.equal(101);
+    expect(size1.height).to.equal(102);
+
+    // Resize.
+    observer.notify([
+      {target: element, contentRect: {width: 201, height: 102}},
+    ]);
+    const size2 = await callbackCaller.next();
+    expect(size2.width).to.equal(201);
+    expect(size2.height).to.equal(102);
+  });
+
+  it('should only observe last change', async () => {
+    const callbackCaller = createCallbackCaller();
+    observeContentSize(element, callbackCaller);
+
+    observer.notify([
+      {target: element, contentRect: {width: 101, height: 102}},
+      {target: element, contentRect: {width: 103, height: 102}},
+    ]);
+    const size1 = await callbackCaller.next();
+    expect(size1.width).to.equal(103);
+    expect(size1.height).to.equal(102);
+    expect(callbackCaller.isEmpty()).to.be.true;
+  });
+
+  it('should observe multiple callbacks', async () => {
+    // First callback.
+    const callbackCaller1 = createCallbackCaller();
+    observeContentSize(element, callbackCaller1);
+    observer.notify([
+      {target: element, contentRect: {width: 101, height: 102}},
+    ]);
+    const size1 = await callbackCaller1.next();
+    expect(size1.width).to.equal(101);
+    expect(size1.height).to.equal(102);
+
+    // Second callback.
+    const callbackCaller2 = createCallbackCaller();
+    observeContentSize(element, callbackCaller2);
+    const size2 = await callbackCaller2.next();
+    expect(size2.width).to.equal(101);
+    expect(size2.height).to.equal(102);
+  });
+
+  it('should unobserve multiple callbacks', async () => {
```

Comment:
nit: can we also have a test that this works for observing elements in different windows

Reply:
Hard, because we don't currently have multi-window fixtures - only one window at a time. However, I'm confirming that I'm using the correct window, because the correct `ResizeObserver` is stubbed.